output/datadog/messagepacker.go

```diff
@@ -0,0 +1,113 @@
+package datadog
+
+import (
+	"bytes"
+
+	"github.com/klauspost/compress/gzip"
+	"github.com/relex/gotils/logger"
+	"github.com/relex/slog-agent/base"
+	"github.com/relex/slog-agent/defs"
+	"github.com/relex/slog-agent/util"
+)
+
+const gzipCompressionLevel = gzip.BestSpeed
+
+// outputChunkMaxDataBytes defines the max uncompressed data size of a LogChunk
+// The value must be lower than the maximum buffer length acceptable by upstream
+var outputChunkMaxDataBytes = 5 * 1024 * 1024
+
+// outputChunkMaxRecords defined the maximum amount of log entries in a single chunk for datadog output
+var outputChunkMaxRecords = 1000
+
+type messagePacker struct {
+	logger              logger.Logger
+	currentChunk        *forwardMessageOpenChunk // current chunk before being made into final message
+	reusedGzipBuffer    *bytes.Buffer            // buffer for gzipWriter for log records
+	reusedMessageBuffer *bytes.Buffer            // buffer for final message
+	tag                 string
+	useCompression      bool
+}
+
+type forwardMessageOpenChunk struct {
+	gzipWriter     *gzip.Writer // gzip writer can be nil if no compression, write to reusedGzipBuffer
+	id             string       // chunk ID
+	numRecords     int          // numbers of collected log records
+	numStreamBytes int          // uncompressed length of written stream data
+}
+
+func NewMessagePacker(parentLogger logger.Logger, tag string) *messagePacker {
+	return &messagePacker{
+		logger:              parentLogger.WithField(defs.LabelComponent, "FluentdForwardMessagePacker"),
+		currentChunk:        nil,
+		reusedGzipBuffer:    bytes.NewBuffer(make([]byte, 0, 1*1024*1024)),
+		reusedMessageBuffer: bytes.NewBuffer(make([]byte, 0, 1*1024*1024)),
+		tag:                 tag,
+		useCompression:      true,
+	}
+}
+
+func (packer *messagePacker) WriteStream(stream base.LogStream) *base.LogChunk {
```

Comment:
Is the code shared with fluentdforward in some way? might be better to extract a compressed chunk writer / splitter which should be a common pattern for all outputs